File: apps/api_keys_router.py
"""
API key management router.

Admin endpoints for generating, listing, and revoking API keys.
Customer endpoint for checking usage/credits.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Header, HTTPException, status

from shared.api_keys import (
    generate_api_key,
    get_next_reset_time,
    get_usage_stats,
    list_all_keys,
    revoke_key,
    validate_api_key,
)
from shared.auth import verify_admin_key
from shared.config import CREDIT_COST_TIERS, DAILY_CREDIT_LIMIT
from shared.schemas import (
    CreditCostInfo,
    GenerateKeyRequest,
    GenerateKeyResponse,
    KeyListResponse,
    UsageResponse,
)

router = APIRouter(prefix="/api/keys", tags=["API Keys"])


# ---------------------------------------------------------------------------
# Admin endpoints (require X-Admin-Key)
# ---------------------------------------------------------------------------

@router.post("/generate", response_model=GenerateKeyResponse)
async def create_key(
    body: GenerateKeyRequest,
    _admin: bool = Depends(verify_admin_key),
) -> GenerateKeyResponse:
    """
    Generate a new API key for a customer.

    The raw key is returned **once** in the response — it cannot be
    retrieved again (only the hash is stored).
    """
    result = generate_api_key(
        owner_email=body.owner_email,
        owner_name=body.owner_name,
        label=body.label,
    )
    return GenerateKeyResponse(**result)


@router.get("", response_model=KeyListResponse)
async def list_keys(
    _admin: bool = Depends(verify_admin_key),
) -> KeyListResponse:
    """List all API keys with their current credit balance (admin only)."""
    keys = list_all_keys()
    return KeyListResponse(total=len(keys), keys=keys)


@router.delete("/{key_id}")
async def delete_key(
    key_id: int,
    _admin: bool = Depends(verify_admin_key),
) -> dict:
    """Revoke an API key (admin only). The key becomes immediately unusable."""
    success = revoke_key(key_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"API key with id={key_id} not found.",
        )
    return {"status": "revoked", "key_id": key_id}


# ---------------------------------------------------------------------------
# Customer endpoints (require X-Api-Key)
# ---------------------------------------------------------------------------

@router.get("/usage", response_model=UsageResponse)
async def check_usage(
    x_api_key: str = Header(..., alias="X-Api-Key"),
) -> UsageResponse:
    """
    Check your credit usage and remaining balance.

    This endpoint does NOT consume credits.
    """
    key_record = validate_api_key(x_api_key)
    if key_record is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Invalid or revoked API key.",
        )

    if key_record["role"] == "owner":
        return UsageResponse(
            is_unlimited=True,
            credits_remaining=-1,
            credits_used_today=0,
            credits_daily_limit=-1,
            resets_at=get_next_reset_time(),
            total_queries_all_time=0,
            total_credits_consumed_all_time=0,
            last_7_days=[],
        )

    # Usage is account-level: pooled across every key this owner_email holds.
    stats = get_usage_stats(key_record["user_id"], key_record["daily_credit_limit"])
    return UsageResponse(**stats)


# ---------------------------------------------------------------------------
# Public endpoints
# ---------------------------------------------------------------------------

@router.get("/pricing", response_model=CreditCostInfo)
async def get_pricing() -> CreditCostInfo:
    """
    Public endpoint showing credit cost tiers and daily limit.

    No authentication required.
    """
    tiers = [
        {"max_characters": max_chars, "credit_cost": cost}
        for max_chars, cost in CREDIT_COST_TIERS
    ]
    return CreditCostInfo(tiers=tiers, daily_limit=DAILY_CREDIT_LIMIT)

File: node-backend/src/routes/apiKeysRouter.ts
/**
 * API key management router.
 *
 * Port of `backend/api_keys_router.py`.
 *
 * Admin endpoints for generating, listing, and revoking API keys. Customer
 * endpoint for checking usage/credits.
 *
 * Mounted at `/api/keys` by `server.ts`.
 */

import { Router } from "express";
import { z } from "zod";

import {
  generateApiKey,
  getNextResetTime,
  getUsageStats,
  listAllKeys,
  revokeKey,
  validateApiKey,
} from "../shared/apiKeys.js";
import { verifyAdminKey } from "../shared/auth.js";
import { CREDIT_COST_TIERS, DAILY_CREDIT_LIMIT } from "../config.js";
import { asyncHandler, forbidden, notFound, parseBody } from "../shared/http.js";

export const router = Router();

const generateKeySchema = z.object({
  owner_email: z.string().min(3).max(255),
  owner_name: z.string().max(255).default(""),
  label: z.string().max(255).default(""),
});

/** The fixed note Pydantic supplied as a field default on GenerateKeyResponse. */
const GENERATED_KEY_MESSAGE =
  "Save this key — it won't be shown again! Credits are shared across " +
  "every key on this account (identified by owner_email), not per-key.";

// ---------------------------------------------------------------------------
// Admin endpoints (require X-Admin-Key)
// ---------------------------------------------------------------------------

/**
 * Generate a new API key for a customer.
 *
 * The raw key is returned **once** in the response — it cannot be retrieved
 * again (only the hash is stored).
 */
router.post(
  "/generate",
  verifyAdminKey,
  asyncHandler(async (req, res) => {
    const body = parseBody(generateKeySchema, req.body);
    const result = await generateApiKey(body.owner_email, body.owner_name, body.label);
    res.json({ ...result, message: GENERATED_KEY_MESSAGE });
  }),
);

/** List all API keys with their current credit balance (admin only). */
router.get(
  "/",
  verifyAdminKey,
  asyncHandler(async (_req, res) => {
    const keys = await listAllKeys();
    res.json({ total: keys.length, keys });
  }),
);

/** Revoke an API key (admin only). The key becomes immediately unusable. */
router.delete(
  "/:keyId",
  verifyAdminKey,
  asyncHandler(async (req, res) => {
    const keyId = req.params.keyId;
    if (!(await revokeKey(keyId))) {
      throw notFound(`API key with id=${keyId} not found.`);
    }
    res.json({ status: "revoked", key_id: keyId });
  }),
);

// ---------------------------------------------------------------------------
// Customer endpoints (require X-Api-Key)
// ---------------------------------------------------------------------------

/**
 * Check your credit usage and remaining balance.
 *
 * This endpoint does NOT consume credits.
 */
router.get(
  "/usage",
  asyncHandler(async (req, res) => {
    const raw = req.headers["x-api-key"];
    const apiKey = (Array.isArray(raw) ? raw[0] : raw) ?? "";

    const keyRecord = apiKey ? await validateApiKey(apiKey) : null;
    if (keyRecord === null) throw forbidden("Invalid or revoked API key.");

    if (keyRecord.role === "owner") {
      res.json({
        is_unlimited: true,
        credits_remaining: -1,
        credits_used_today: 0,
        credits_daily_limit: -1,
        resets_at: getNextResetTime(),
        total_queries_all_time: 0,
        total_credits_consumed_all_time: 0,
        last_7_days: [],
      });
      return;
    }

    // Usage is account-level: pooled across every key this owner_email holds.
    const stats = await getUsageStats(keyRecord.user_id, keyRecord.daily_credit_limit);
    res.json({ is_unlimited: false, ...stats });
  }),
);

// ---------------------------------------------------------------------------
// Public endpoints
// ---------------------------------------------------------------------------

/**
 * Public endpoint showing credit cost tiers and daily limit.
 *
 * No authentication required.
 */
router.get(
  "/pricing",
  asyncHandler(async (_req, res) => {
    res.json({
      tiers: CREDIT_COST_TIERS.map(([maxChars, cost]) => ({
        max_characters: maxChars,
        credit_cost: cost,
      })),
      daily_limit: DAILY_CREDIT_LIMIT,
    });
  }),
);
